src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -264,18 +265,31 @@ public void SetAll(bool value)
         ** Exceptions: ArgumentException if value == null or
         **             value.Length != this.Length.
         =========================================================================*/
-        public BitArray And(BitArray value)
+        public unsafe BitArray And(BitArray value)
         {
             if (value == null)
                 throw new ArgumentNullException(nameof(value));
             if (Length != value.Length)
                 throw new ArgumentException(SR.Arg_ArrayLengthsDiffer);
 
-            for (int i = 0; i < m_array.Length; i++)
+            int i = 0;
```

Comment:
Can we get rid of some of the regression for small numbers by special-casing and manually unrolling the `1 through 8` case?

Reply:
Worth a shot :)

Reply:
Something like:
```
var count = m_array.Length;

switch (count)
{
    case 7: m_array[7] &= value.m_array[7]; goto case 6;
    case 6: m_array[6] &= value.m_array[6]; goto case 5;
    case 5: m_array[5] &= value.m_array[5]; goto case 4;
    case 4: m_array[4] &= value.m_array[4]; goto case 3;
    case 3: m_array[3] &= value.m_array[3]; goto case 2;
    case 2: m_array[2] &= value.m_array[2]; goto case 1;
    case 1: m_array[1] &= value.m_array[1]; goto _done;
}
```

Reply:
or maybe something like:
```
var count = m_array.Length;

switch (count)
{
    case 3: m_array[3] &= value.m_array[3]; goto case 2;
    case 2: m_array[2] &= value.m_array[2]; goto case 1;
    case 1: m_array[1] &= value.m_array[1]; goto _done;

    case 7: m_array[7] &= value.m_array[7]; goto case 6;
    case 6: m_array[6] &= value.m_array[6]; goto case 5;
    case 5: m_array[5] &= value.m_array[5]; goto case 4;
    case 4: result = vector & vector; goto _done;

    case 8: result = vector & vector; goto case 4;
}
```

Reply:
The performance of the switch vs. for-loop for patterns like these depends a lot on the predictability of `count`. If it is predictable (e.g. always the same), `switch` is somewhat better. If if is non-predictable (e.g. random), `switch` tends to be much worse.

I think the current code is better because of it is smaller.

Reply:
> Does the JIT not just emit a jump table for small switch cases like this?

The threshold is much less than 7.

Reply:
> Does the JIT not just emit a jump table for small switch cases like this? 

Yes, it does. The jump tables work great in micro-benchmarks, but not as good in real world workloads. The indirect jumps tend to get poorly predicted.

Reply:
(and since the jump is unconditional, there should be no miss-prediction)

Reply:
The processor needs to know where the branch is going to go to keep decoding instruction stream. For indirect jumps, it cannot even guess taken/not-taken like for conditional jumps when it has no data for the given branch in the predictor. It needs to wait for the execution to catch up.

Reply:
> There should only be a single jump generated....
> (and since the jump is unconditional, there should be no miss-prediction)

Agree with @jkotas, the branch prediction is not related how many jumps you have or unconditional/conditional jump. In your example, `jmp r8` is an indirect jump but CPU cannot predict its targets if the data (`DWORD PTR $LN14@And[rcx+rax*4]`) does not have obvious patterns.

> The indirect jumps tend to get poorly predicted.

Another reason for poorly predictability is `switch` only has one dispatch node for all `case`s, so that the branch history table has to remember recently seen targets for all cases. But `if` sequences (or other equivalents, like threaded code ) has a dispatch node for each "case", so each branch history table just needs to remember its own recent targets, which would have much better branch prediction rate.

Reply:
Just in case, I tried to unroll first 3 values using `if` and `switch` (see https://gist.github.com/EgorBo/66fad0c6b50775ec0ec022a14437bd77) and got these results on my machine: https://user-images.githubusercontent.com/523221/49330163-00af6c80-f59c-11e8-8733-9a4707fa0ea7.png